packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp

```diff
@@ -0,0 +1,499 @@
+#include "Tpetra_Core.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Tpetra_Import.hpp"
+#include "Tpetra_FEMultiVector.hpp"
+#include "Tpetra_MultiVector_decl.hpp"
+#include "Tpetra_MultiVector_def.hpp"
+#include "Zoltan2_VtxLabel.hpp"
+
+#include <string>
+#include <sstream>
+#include <iostream>
+#include <queue>
+#include <vector>
+
+#ifndef ZOLTAN2_ICEPROP_
+#define ZOLTAN2_ICEPROP_
+
+namespace Zoltan2{
+template<typename MAP>
+class iceSheetPropagation {
+public:
+
+  //typedefs for the FEMultiVector
+  typedef typename MAP::local_ordinal_type lno_t;
+  typedef typename MAP::global_ordinal_type gno_t;
+  typedef IcePropVtxLabel<lno_t,gno_t> scalar_t;
+  typedef IcePropVtxLabel<lno_t,gno_t> label_t;
+  typedef Tpetra::FEMultiVector<scalar_t,lno_t, gno_t> femv_t;	
+  
+  std::queue<lno_t> artQueue;
+  std::queue<lno_t> regQueue;
+ 	
+  //Constructor creates FEMultiVector of IcePropVertexLabels,
+  //and initializes the labels according to the input grounding information.
+  //It also sets the is_art flag on each label if the vertex is a potential articulation point.
+  iceSheetPropagation(const Teuchos::RCP<const Teuchos::Comm<int> > &comm_, Teuchos::RCP<const MAP> mapOwned_, Teuchos::RCP<const MAP> mapWithCopies_, icePropGraph<lno_t>* g_,int* boundary_flags, bool* grounding_flags,int localOwned,int localCopy):
+    me(comm_->getRank()), np(comm_->getSize()),
+    nLocalOwned(localOwned), nLocalCopy(localCopy),
+    nVec(1), comm(comm_),g(g_),mapOwned(mapOwned_),
+    mapWithCopies(mapWithCopies_)
+  {
+    typedef Tpetra::Import<lno_t, gno_t> import_t;
+    Teuchos::RCP<import_t> importer = rcp(new import_t(mapOwned, mapWithCopies));
+    femv = rcp(new femv_t(mapOwned, importer, nVec, true)); 
+    femv->beginFill();
+    //set the member variable that stores femv->getData(0);
+    femvData = femv->getData(0);
+    for(lno_t i = 0; i < nLocalOwned + nLocalCopy; i++){
```

Comment:
Why does this for-loop loop over the ghost entries (nLocalCopy)?
The boundary flag test (line 51) assures that only owned vertices are marked as articulation points.
The grounding flags test (line 54) will be false for all ghosts (see comment in Zoltan2_IceSheet.hpp).
So nothing will be done for ghost vertices here, right?
Should this loop go over only owned vertices?
Or should grounding_flag info in Zoltan2_IceSheet.hpp be updated with ghosts' grounding info?

Reply:
This loop initializes the label datastructure for all vertices, so I do need to loop through ghost entries. However, I do not need to use the grounding information for ghost entries, just set the global ID for future reference. I have changed this as well so the smaller grounding array will not cause issues here. Tests pass.